doc/tutorials/platformer/step_3.md

```diff
@@ -0,0 +1,458 @@
+    );
+    add(_ember);
+  }
+```
+
+We simply are taking the width of the game screen, divide that by 640 (10 blocks in a segment times
+64 pixels wide for each block), and round that up.  As we only defined 5 segments total, we need to
+restrict that integer from 0 to the length of the segments list in case the user has a really wide
+screen.  Then we simply loop through the number of `segmentsToLoad` and call `loadGameSegments` with
+the integer to load and then calculate the offset.
+
+Additionally, I have moved the Ember-related code from the `onLoad` method to our new
+`initializeGame` method.  This means I can now make the call in `onLoad` to `initializeGame` such
+as:
+
+```dart
+@override
+  Future<void> onLoad() async {
+    await images.loadAll([
+      'block.png',
+      'ember.png',
+      'ground.png',
+      'heart_half.png',
+      'heart.png',
+      'star.png',
+      'water_enemy.png',
+    ]);
+    initializeGame();
+  }
+```
+
+At this point, you probably have errors for all the object classes and the enemy class, but don't
+worry, we will solve those right now.
+
+
+### The Platform Block
+
+One of the easiest blocks to start with is the Platform Block.  There are two things that we need to
+develop beyond getting the sprite to be displayed; that is, we need to place it in the correct
+position and as Ember moves across the screen, we need to remove the blocks once they are off the
+screen.  In Ember Quest, the player can only move forward, so this will keep the game lightweight as
+it's an infinite level.
+
+Open the `lib/objects/platform_block.dart` file and add the following code:
+
+```dart
+import 'package:flame/collisions.dart';
+import 'package:flame/components.dart';
+
+import '../ember_quest.dart';
+
+class PlatformBlock extends SpriteComponent
+    with HasGameRef<EmberQuestGame> {
+  final Vector2 gridPosition;
+  double xOffset;
+
+  PlatformBlock({
+    required this.gridPosition,
+    required this.xOffset,
+  }) : super(size: Vector2.all(64), anchor: Anchor.bottomLeft);
+
+  @override
+  Future<void> onLoad() async {
+  }
+
+  @override
+  void update(double dt) {
+    super.update(dt);
+  }
+}
+```
+
+We are going to extend the Flame `SpriteComponent` and we will need the `HasGameRef` mixin to access
+our game class just like we did before.  We are starting with the empty `onLoad` and `update`
+methods and we will begin adding code to create the functionality that is necessary for the game.
+
+The secret to any gaming engine is the game loop.  This is an infinite loop that calls all the
+objects in your game so you can provide updates.  The `update` method is the hook into this and it
+uses a `double dt` to pass to your method the amount of time in microseconds since it was last
+called.  This `dt` variable then allows you to calculate how far your component needs to move
+on-screen.  
+
+All components in our game will need to move at the same speed, so to do this, open
```

Comment:
Wait, so the entire world "flies by", while the main character simply hops in place? Like in the TRex game? This may need to be explained in more details, because it's a rather non-trivial architectural detail.

Also, for a more advanced platformer it would probably be better to use more natural abstractions of a static world and a camera that follows the main character.

Reply:
Oh is that the case? I also missed the fact that it is the world flying by and not the player moving in it. Would it be hard to convert the code to do that instead @munsterlander? Because there are already quite some trex-style tutorials around.

Reply:
This is not the case. The player is moving and this behaves just like the original Mario. You can move forward and back in the world that is on screen. When you get to a set distance to the right, the next world segment is loaded.

Reply:
That being said, perhaps the current approach makes it easier to generate new parts of the world on-demand? It might be good to at least add some paragraph discussing pros and cons of different approaches.

Reply:
Yes. I would contend traditional platformers would have levels like in Mario and normally I would use tiled to do that, but we wanted to make this flame only.  I could use the segment manager to make like an end level that then wipes the screen into a reset or pop an overlay or whatever but I was afraid it would get too complex. I will definitely add a paragraph to explain the difference.

Reply:
Paragraphs added using the note annotation.